cmd/starter/c/starter.c

```diff
@@ -1002,182 +959,235 @@ __attribute__((constructor)) static void init(void) {
             action.sa_sigaction = (void *)&do_exit;
             action.sa_flags = SA_SIGINFO|SA_RESTART;
 
-            close(master_socket[0]);
-            close(master_socket[1]);
-
             sigemptyset(&usrmask);
             sigaddset(&usrmask, SIGUSR1);
-            sigaddset(&usrmask, SIGUSR2);
 
-            if (sigprocmask(SIG_SETMASK, &usrmask, NULL) == -1) {
+            if ( sigprocmask(SIG_SETMASK, &usrmask, NULL) == -1 ) {
                 fatalf("Blocked signals error: %s\n", strerror(errno));
             }
-            if (sigaction(SIGUSR2, &action, NULL) < 0) {
-                fatalf("Failed to install signal handler for SIGUSR2\n");
-            }
-            if (sigaction(SIGUSR1, &action, NULL) < 0) {
+            /* master process will send SIGUSR1 to detach successfully */
+            if ( sigaction(SIGUSR1, &action, NULL) < 0 ) {
                 fatalf("Failed to install signal handler for SIGUSR1\n");
             }
-            if (sigprocmask(SIG_UNBLOCK, &usrmask, NULL) == -1) {
+            if ( sigprocmask(SIG_UNBLOCK, &usrmask, NULL) == -1 ) {
                 fatalf("Unblock signals error: %s\n", strerror(errno));
             }
-            while ( waitpid(forked, &status, 0) <= 0 ) {
-                continue;
-            }
-            exit_with_status("instance", status);
+            /* loop until master process exits with error */
+            wait_child("instance", process, 1);
```

Comment:
I originally had some confusion regarding the intention of this function call here, and whether it would ever return and unintentionally continue to execute `init()`. After a slack conversation with @cclerget I think I understand what's going on, which I'll try to recap:

- When requested to run as an instance, we `fork()` to create a parent/child
- The child process calls `umask(0)` and then continues through `init()`
- The parent process exists to monitor if `master` process succeeded
- Parent process creates a handler for `SIGUSR1`. Master process will send `SIGUSR1` if master stage is successful. The signal handler declared will then essentially `exit(0)` as it was successful.
- `wait_child("instance", process, 1);` is called. The third arg of `wait_child` signifies if the function should exit after waiting on the child has finished.
- There is thus no case where `wait_child` call should return and allow the parent process to continue through `init()` which was my initial worry.

I would suggest that to increase clarity and intent, we do this:

```
wait_child("instance", process, 1);

/* we should never return from the previous wait_child call */
exit(1);
```

This way we are signaling our intent in code & comment